polarCoordinateTraversal_dotprod.m

```diff
@@ -0,0 +1,198 @@
+    plot(ray_end_x, ray_end_y, 'k.', 'MarkerSize', 10);
+    plot(ray_start_x, ray_start_y, 'k.', 'MarkerSize', 10);
+    quiver(ray_start_x, ray_start_y, ray_direction_x, ray_direction_y, t_end - t_begin, 'LineWidth', 1.5);
+    
+    % Draw the axis.
+    axis tight;
+    xlim([min_bound_x, max_bound_x]);
+    ylim([min_bound_y, max_bound_y]);
+    xlabel('x');
+    ylabel('y');
+    grid on;
+    
+    % Draw the radial sections.
+    current_max_radius = circle_max_radius;
+    delta_radius = circle_max_radius / num_radial_sections;
+    for k = 1:num_radial_sections
+        viscircles(circle_center, current_max_radius, 'LineStyle', '--', 'Color', '#7E2F8E', 'LineWidth', 1);
+        current_max_radius = current_max_radius - delta_radius;
+    end
+    
+    % Draw the angular sections.
+    N = num_angular_sections;
+    section = 2 * pi / num_angular_sections;
+    for ii = 1:N
+        t = linspace(section * (ii - 1), section * (ii));
+        x = circle_max_radius*cos(t) + circle_center_x;
+        y = circle_max_radius*sin(t) + circle_center_y;
+        x = [x circle_center_x x(1)];
+        y = [y circle_center_y y(1)];
+        line(x, y, 'LineStyle', '--', 'Color', '#7E2F8E', 'LineWidth', 0.5);
+    end
+end
+
+% INITIALIZATION PHASE
+% Where is the ray at t_begin?
+p = ray_origin + t_begin.*ray_direction;
+ray_circle_vector = [circle_center(1) - p(1); circle_center(2) - p(2)]';
+% Find the radial shell containing the ray at t_begin
+r = delta_radius;
+while ((ray_circle_vector(1)^2 + ray_circle_vector(2)^2 > r^2) && ...
+        r < circle_max_radius)
+    r = r + delta_radius        ;
+end
+% Find the intersection times for the ray and the radial shell containing the ray
+% at t_begin
+ray_unit_vector = 1/sqrt(ray_direction(1)^2 + ray_direction(2)^2)...
+    .* [ray_direction(1);  ray_direction(2)]';
+v = dot(ray_circle_vector,ray_unit_vector);
+discr = r^2 - (dot(ray_circle_vector,ray_circle_vector) - v^2);
+d = sqrt(discr);
+p1 = ray_origin + (v-d).*ray_unit_vector;
+p2 = ray_origin + (v+d).*ray_unit_vector;
+t1 = (p1(1)-ray_origin(1))/ray_direction(1);
+t2 = (p2(1)-ray_origin(2))/ray_direction(2);
+% The ray may not intersect the grid
+% at all; in particular if the ray is outside the grid at t_begin.
+if (t1 < t_begin && t2 < t_begin )
+fprintf("\nray doesn't intersect polar grid for t_begin!")
+return;
+end
+% If there is a ray/shell intersection, then set the radial voxel ID. 
+current_voxel_ID_r = 1 + (circle_max_radius - r)/delta_radius
+if verbose
+    text(p1(1), p1(2), 'POI_r');
+    fprintf('RADIAL HIT INIT.\n');
+end
+
+% II. Calculate Voxel ID Theta.
+current_voxel_ID_theta = floor(atan2(ray_start_y - circle_center_y, ray_start_x - circle_center_x) * num_angular_sections / (2 * pi));
+if current_voxel_ID_theta < 0
+    current_voxel_ID_theta = num_angular_sections + current_voxel_ID_theta;
+end
+angular_voxels = [current_voxel_ID_theta];
+radial_voxels = [current_voxel_ID_r];
+
+% III. TRAVERSAL PHASE
+tol = 10^-6;
```

Comment:
Instead of using our own tol, do we just want to use MATLAB's ```eps```?
See: https://www.mathworks.com/help/matlab/ref/eps.html

Reply:
OK. Didn't know about eps. We can use that instead of defining our own tolerance.